### PR TITLE
Fast intrinsics and fastmath

### PR DESCRIPTION
**Pull request by maleadt**

### CODE REVIEW

**Comment by vchuravy** [comment]
Is there an actual performance difference with use the fast methods? Also fastmath is much broader than just using the fast methods

**Comment by maleadt** [comment]
I just remembered @jrevels mentioning that comparing against an as much as default as possible TF set-up would be preferable. And sure, fastmath is much more, but also with these simple benchmarks?

**Comment by maleadt** [comment]
Looking into this, it seemed like `@fastmath` had a negative effect on generated code quality. With the following MWE, `div` and `mul` get converted to `neg`+`div`+`fma`+`neg`, with 2 additional registers used:

```julia
using CUDAnative

struct Dual{T}
    x::T
    y::T
end

@inline function regular(a, b)
    c = 1 / (1 + (-a))
    d = (-b) * (- c / (1 + (-a)))
    return Dual(c, d)
end

function fast(a, b)
    c = @fastmath 1 / (1 + (-a))
    d = @fastmath (-b) * (- c / (1 + (-a)))
    return Dual(c, d)
end

CUDAnative.code_ptx(regular, NTuple{2,Float32})

CUDAnative.code_ptx(fast, NTuple{2,Float32})
```


Reduced from:

```julia
using ForwardDiff, CUDAnative

regular(x) = 1 / (1 + (-x))

fast(x) = @fastmath 1 / (1 + (-x))


CUDAnative.code_ptx(regular, Tuple{ForwardDiff.Dual{Nothing,Float32,1}})

CUDAnative.code_ptx(fast, Tuple{ForwardDiff.Dual{Nothing,Float32,1}})
```


Reduced from:

```julia
using ForwardDiff, DiffRules, CUDAnative

DiffRules.@define_diffrule CUDAnative.exp_fast(x) = :(CUDAnative.exp_fast($x))
DiffRules.@define_diffrule CUDAnative.tanh(x) = :(1 - CUDAnative.tanh($x)^2)

@eval $(ForwardDiff.unary_dual_definition(:CUDAnative, :exp_fast))
@eval $(ForwardDiff.unary_dual_definition(:CUDAnative, :tanh))


sigm(x) = 1 / (1 + CUDAnative.exp_fast(-x))
tanh(x) = CUDAnative.tanh(x)

function regular(z, zb, c, f, i, g)
    if z == 1.0f0
        return sigm(i) * tanh(g)
    elseif zb == 0.0f0
        return c
    else
        return sigm(f) * c + sigm(i) * tanh(g)
    end
end


fast_sigm(x) = @fastmath 1 / (1 + CUDAnative.exp_fast(-x))
fast_tanh(x) = @fastmath CUDAnative.tanh(x)

function fast(z, zb, c, f, i, g)
    if z == 1.0f0
        return fast_sigm(i) * fast_tanh(g)
    elseif zb == 0.0f0
        return c
    else
        return fast_sigm(f) * c + fast_sigm(i) * fast_tanh(g)
    end
end


dual_t = ForwardDiff.Dual{Nothing,Float32,5}
tt = NTuple{6,dual_t}

CUDAnative.code_ptx(regular, tt)

CUDAnative.code_ptx(fast, tt)
```


Reduced from MixedModeBroadcastAD.


With this last repro, register pressure regresses from 200 to 238 f32 registers... Can't be faster, right?

```julia
# code from above

dual = ForwardDiff.Dual(1f0,2f0,3f0,4f0,5f0,6f0)

@inline sink(i::Float32) =
    Base.llvmcall("""%slot = alloca float
                     store volatile float %0, float* %slot
                     %value = load volatile float, float* %slot
                     ret float %value""", Float32, Tuple{Float32}, i)

function kernel(fun, z, zb, c, f, i, g)
    dual = fun(z, zb, c, f, i, g)
    sink(dual.value)
    sink(dual.partials[1])
    sink(dual.partials[2])
    sink(dual.partials[3])
    sink(dual.partials[4])
    sink(dual.partials[5])
end

@cuda kernel(regular, dual, dual, dual, dual, dual, dual)

@cuda kernel(fast, dual, dual, dual, dual, dual, dual)
```

Apparently LLVM is better at compiling than I am; this code is actually faster. I guess the most important factor is the increase in (virtual) register pressure mostly disappearing when compiling to actual machine code (even with high arity). Hadn't considered PTX being SSA-like, and hence register pressure not meaning much.

TL;DR: `@fastmath` matters, non-tfstyle HMLSTM improves couple of % despite slightly higher register pressure. I'll do some proper benchmarks tomorrow.

**Comment by maleadt** [comment]
XLA with/without fastmath doesn't seem to matter:

```
julia> df = process("test/reference"); df[df[:system].==:python, :]
3×13 DataFrame
│ Row │ system │ TFstyle │ arity   │ dims │ duration        │ kernel_count │ kernel_registers │ kernel_duration  │ kernel_occupancy │ memcpy_count │ memcpy_duration │ api_count │ api_duration    │
├─────┼────────┼─────────┼─────────┼──────┼─────────────────┼──────────────┼──────────────────┼──────────────────┼──────────────────┼──────────────┼─────────────────┼───────────┼─────────────────┤
│ 1   │ python │ missing │ missing │ 512  │ 1673.28±450.823 │ 7            │ 112              │ 73.7291±0.217993 │ 0.106902         │ 5            │ 325.519±3.68087 │ 17        │ 261.191±28.1929 │
│ 2   │ python │ missing │ missing │ 1024 │ 2563.14±91.8891 │ 7            │ 112              │ 221.973±0.908082 │ 0.144441         │ 5            │ 1287.14±9.94711 │ 17        │ 276.748±26.7941 │
│ 3   │ python │ missing │ missing │ 2048 │ 6919.3±66.9751  │ 7            │ 112              │ 818.152±1.62282  │ 0.14358          │ 5            │ 5121.26±21.1479 │ 17        │ 263.103±25.2284 │

julia> df = process("test/fastmath"); df[df[:system].==:python, :]
3×13 DataFrame
│ Row │ system │ TFstyle │ arity   │ dims │ duration        │ kernel_count │ kernel_registers │ kernel_duration  │ kernel_occupancy │ memcpy_count │ memcpy_duration │ api_count │ api_duration    │
├─────┼────────┼─────────┼─────────┼──────┼─────────────────┼──────────────┼──────────────────┼──────────────────┼──────────────────┼──────────────┼─────────────────┼───────────┼─────────────────┤
│ 1   │ python │ missing │ missing │ 512  │ 1616.55±182.195 │ 7            │ 114              │ 74.6523±0.511521 │ 0.107373         │ 5            │ 327.406±4.04173 │ 17        │ 283.913±32.2684 │
│ 2   │ python │ missing │ missing │ 1024 │ 2458.5±99.1087  │ 7            │ 114              │ 224.507±0.947644 │ 0.14325          │ 5            │ 1286.2±8.2451   │ 17        │ 260.849±35.7714 │
│ 3   │ python │ missing │ missing │ 2048 │ 6915.36±114.78  │ 7            │ 114              │ 817.926±1.00716  │ 0.14672          │ 5            │ 5114.41±12.6215 │ 17        │ 293.132±56.5484 │
```

But with `@fastmath`, there's a significant improvement!

```
julia> df = process("test/reference"); df = df[.!ismissing.(df[:TFstyle]), :]; df[df[:TFstyle] .== false, :]
3×13 DataFrame
│ Row │ system │ TFstyle │ arity   │ dims │ duration        │ kernel_count │ kernel_registers │ kernel_duration  │ kernel_occupancy │ memcpy_count │ memcpy_duration │ api_count │ api_duration    │
├─────┼────────┼─────────┼─────────┼──────┼─────────────────┼──────────────┼──────────────────┼──────────────────┼──────────────────┼──────────────┼─────────────────┼───────────┼─────────────────┤
│ 1   │ julia  │ false   │ missing │ 512  │ 124.342±19.2416 │ 1            │ 39               │ 19.1526±0.234234 │ 0.589884         │ 0            │ 0.0±0.0         │ 4         │ 61.0391±9.46079 │
│ 2   │ julia  │ false   │ missing │ 1024 │ 174.554±33.7206 │ 1            │ 39               │ 58.5847±0.540268 │ 0.659677         │ 0            │ 0.0±0.0         │ 4         │ 68.6577±28.978  │
│ 3   │ julia  │ false   │ missing │ 2048 │ 307.697±13.6793 │ 1            │ 39               │ 206.993±0.373724 │ 0.680399         │ 0            │ 0.0±0.0         │ 4         │ 62.0093±11.6952 │

julia> df = process("test/fastmath"); df = df[.!ismissing.(df[:TFstyle]), :]; df[df[:TFstyle] .== false, :]
3×13 DataFrame
│ Row │ system │ TFstyle │ arity   │ dims │ duration        │ kernel_count │ kernel_registers │ kernel_duration  │ kernel_occupancy │ memcpy_count │ memcpy_duration │ api_count │ api_duration    │
├─────┼────────┼─────────┼─────────┼──────┼─────────────────┼──────────────┼──────────────────┼──────────────────┼──────────────────┼──────────────┼─────────────────┼───────────┼─────────────────┤
│ 1   │ julia  │ false   │ missing │ 512  │ 128.108±13.9394 │ 1            │ 35               │ 18.2266±0.299602 │ 0.593296         │ 0            │ 0.0±0.0         │ 4         │ 62.616±10.157   │
│ 2   │ julia  │ false   │ missing │ 1024 │ 154.384±13.2192 │ 1            │ 35               │ 53.2898±0.519898 │ 0.65318          │ 0            │ 0.0±0.0         │ 4         │ 62.004±10.618   │
│ 3   │ julia  │ false   │ missing │ 2048 │ 294.501±11.8155 │ 1            │ 35               │ 190.069±0.623129 │ 0.672016         │ 0            │ 0.0±0.0         │ 4         │ 64.8907±10.6567 │
```

Strangely, register pressure is lower. Even though my analysis last week seemed to imply the inverse.
Let me look into this a little more and merge it then.

TF-style is unaffected by `@fastmath`:

```
julia> df = process("test/reference"); df = df[.!ismissing.(df[:TFstyle]), :]; df[df[:TFstyle] .== true, :]
3×13 DataFrame
│ Row │ system │ TFstyle │ arity   │ dims │ duration        │ kernel_count │ kernel_registers │ kernel_duration  │ kernel_occupancy │ memcpy_count │ memcpy_duration │ api_count │ api_duration    │
├─────┼────────┼─────────┼─────────┼──────┼─────────────────┼──────────────┼──────────────────┼──────────────────┼──────────────────┼──────────────┼─────────────────┼───────────┼─────────────────┤
│ 1   │ julia  │ true    │ missing │ 512  │ 300.985±27.8348 │ 7            │ 164              │ 57.4156±0.31708  │ 0.663576         │ 0            │ 0.0±0.0         │ 16        │ 148.679±19.3021 │
│ 2   │ julia  │ true    │ missing │ 1024 │ 318.969±26.9049 │ 7            │ 164              │ 139.602±0.443393 │ 0.823846         │ 0            │ 0.0±0.0         │ 16        │ 162.545±19.5678 │
│ 3   │ julia  │ true    │ missing │ 2048 │ 581.96±8.95881  │ 7            │ 164              │ 447.968±0.381366 │ 0.869699         │ 0            │ 0.0±0.0         │ 16        │ 169.323±13.5523 │

julia> df = process("test/fastmath"); df = df[.!ismissing.(df[:TFstyle]), :]; df[df[:TFstyle] .== true, :]
3×13 DataFrame
│ Row │ system │ TFstyle │ arity   │ dims │ duration        │ kernel_count │ kernel_registers │ kernel_duration  │ kernel_occupancy │ memcpy_count │ memcpy_duration │ api_count │ api_duration    │
├─────┼────────┼─────────┼─────────┼──────┼─────────────────┼──────────────┼──────────────────┼──────────────────┼──────────────────┼──────────────┼─────────────────┼───────────┼─────────────────┤
│ 1   │ julia  │ true    │ missing │ 512  │ 309.676±18.2203 │ 7            │ 164              │ 57.805±0.321641  │ 0.662544         │ 0            │ 0.0±0.0         │ 16        │ 152.678±11.8521 │
│ 2   │ julia  │ true    │ missing │ 1024 │ 350.597±25.2671 │ 7            │ 164              │ 139.652±0.374709 │ 0.823881         │ 0            │ 0.0±0.0         │ 16        │ 183.266±17.6262 │
│ 3   │ julia  │ true    │ missing │ 2048 │ 582.565±32.7269 │ 7            │ 164              │ 447.988±0.38349  │ 0.869685         │ 0            │ 0.0±0.0         │ 16        │ 180.204±44.7475 │
```
